contents/blog/null-handling-hogql.md

```diff
@@ -0,0 +1,82 @@
+---
+title: "Adventures in null handling: Null money, null problems"
+date: 2023-06-14
+author: ["ian-vanagas"]
+showTitle: true
+rootpage: /blog
+sidebar: Blog
+hideAnchor: true
+featuredImage: ../images/blog/orange-blog-image.jpg
+featuredImageType: full
+category: Engineering
+---
+
+The value (or non-value) null can be the bane of a programmer’s existence. Null can work as expected in one language and context, then act in completely unexpected ways in another language or context.
+
+We recently had challenges with null during the [release of HogQL](/blog/introducing-hogql). Our engineers Marius and Michael realized null values for properties were leading to inaccurate results and customer confusion. This post covers their efforts to figure out and implement a solution for this.
+
+## The problem with null in HogQL
+
+The issue with null in HogQL was that SQL is strict about nulls. If any value in an expression is null, then the entire expression is null. Here are two HogQL expressions showcasing why this is a problem:
+
+1. `sumIf(properties.money, event = 'no problems') + sumIf(properties.money - 10, event = 'some problems')` returns null if either of the sides of the equation are null. This  happens if there are no matching events (named `no problems` or `some problems`) or the property (`money`) is null.
+2. `concat('money: ', properties.money)` returns null if there's no `money` property, though users might expect either `'money: '`  or `'money: null'`
+
+PostHog’s existing solutions for handling null did not work with the full access HogQL provides. Specifically:
+
+- ClickHouse skips null during aggregation, but users could want to concat or check null values with HogQL.
+- PostHog previously treated nulls as strings, but this could include null values in counts when they shouldn’t be.
+
+At this stage in HogQL’s development, if a value could be null, users must handle it themselves. For example, users must wrap each `sumIf` function in the first example above in a `nullIf(x, 0)` like this:
```

Comment:
I feel there is some backgruond context missing. Specifically:

- What is HogQL?
- Tech stack e.g., clickhouse etc. 

It was a bit confusing to jump straight into things

Reply:
Added a half sentence on both of these.